homeassistant/util/dt.py

```diff
@@ -337,11 +344,32 @@ def _lower_bound(arr: list[int], cmp: int) -> int | None:
     # For example: if triggering on 2:30 and now is 28.10.2018 2:30 (in DST)
     # we should trigger next on 28.10.2018 2:30 (out of DST), but our
     # algorithm above would produce 29.10.2018 2:30 (out of DST)
-    if tz.datetime_ambiguous(now):
+    if datetime_ambiguous(now):
         check_result = find_next_time_expression_time(
             now + _dst_offset_diff(now), seconds, minutes, hours
         )
-        if tz.datetime_ambiguous(check_result):
+        if datetime_ambiguous(check_result):
             return check_result
 
     return result
+
+
+def datetime_exists(dattim: dt.datetime) -> bool:
+    """Check if a datetime exists."""
+    assert dattim.tzinfo is not None
+    original_tzinfo = dattim.tzinfo
+    # Check if we can round trip to UTC
+    return dattim.replace(tzinfo=None) == dattim.astimezone(UTC).astimezone(
+        original_tzinfo
+    ).replace(tzinfo=None)
+
+
+def datetime_ambiguous(dattim: dt.datetime) -> bool:
+    """Check whether a datetime is ambiguous."""
+    assert dattim.tzinfo is not None
+    non_folded = dattim.replace(fold=0)
+    folded = dattim.replace(fold=1)
+    return not (
```

Comment:
This will return True for imaginary times as well:

```
>>> datetime_ambiguous(dt.datetime(2020, 3, 8, 2, 30, tzinfo=zoneinfo.ZoneInfo("America/New_York"))
True
```

I think you can write this function this way:

```python
opposite_fold = dattim.replace(fold=not dattim.fold)
return datetime_exists(dattim) and dattim.utcoffset() != opposite_fold.utcoffset()
```

Though in the current usage, `datetime_ambiguous` is only ever called if `datetime_exists` is `False` anyway, so maybe you don't care (though it may be worth a comment if not an assertion).

Reply:
Thanks!